### PR TITLE
fix(postgres): prevent crash if postgres connection emit multiple errors

### PR DESCRIPTION
**Pull request by Hackatosh**
## Pull Request Checklist

- [ ] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [ ] Did you update the typescript typings accordingly (if applicable)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

We recently observed multiple crashes due to an unhandled "server conn crashed?" exception originating from the pg library. As best as I've been able to tell, this shouldn't be possible and should have been fixed in this previous PR : https://github.com/sequelize/sequelize/pull/14731

I think that an error might be able to slip up during the connection without the error listener being properly attached or that error can be fired multiple times and is only handled once.
My fix is to attach the listener before trying to connect to the database. Also I change back the .once handler to a .on (this was changed in the PR 14731) to be able to catch multiple errors.

### CODE REVIEW

**Comment by evanrittenhouse** [comment]
LGTM - will wait for @ephys or @WikiRik to confirm, since they may have additional insight into to the Postgres error  emission logic

**Review by evanrittenhouse — DISMISSED** [review]

File: packages/core/src/dialects/postgres/connection-manager.ts [hunk 1]
```diff
@@ -193,9 +193,9 @@ export class PostgresConnectionManager extends AbstractConnectionManager<PgConne
     });
 
     // Don't let a Postgres restart (or error) to take down the whole app
-    connection.once('error', error => {
+    connection.on('error', error => {
       connection._invalid = true;
-      debug(`connection error ${error.code || error.message}`);
+      debug(`connection error ${(error as unknown as { code: string }).code || error.message}`);
```
Comment (WikiRik) [inline packages/core/src/dialects/postgres/connection-manager.ts:198]: Is this still needed here?

Comment (WikiRik) [reply packages/core/src/dialects/postgres/connection-manager.ts:198]: Update; it is. When using once it sees `error` as `any`, but using `on` it is typed as an `Error` which does not have a code.

```suggestion
      debug(`connection error ${error.code || error.message}`);
```

File: packages/core/src/dialects/postgres/connection-manager.ts [hunk 2]
```diff
@@ -193,9 +193,9 @@ export class PostgresConnectionManager extends AbstractConnectionManager<PgConne
     });
 
     // Don't let a Postgres restart (or error) to take down the whole app
-    connection.once('error', error => {
+    connection.on('error', error => {
```
Comment (WikiRik) [inline packages/core/src/dialects/postgres/connection-manager.ts:196]:
```suggestion
    connection.on('error', (error: any) => {
```

**Review by WikiRik — COMMENTED** [review]
Suggestion to make the casing of the type a bit nicer

**Comment by Hackatosh** [comment]
Thanks for your input @WikiRik, I have updated the PR

**Review by WikiRik — APPROVED** [review]
Looks good to me, but I'll give @ephys some time to look at the typing change before I merge

**Comment by Hackatosh** [comment]
Hello ! Any news on this PR ? @WikiRik @ephys

**Pull request merged**